### DISCUSSION TITLE
Unable to silence logging messages

### DISCUSSION THREAD

**Discussion by AlexKurek** [pallets/flask#5443] [Q&A]
Im getting ~15 of:

> 127.0.0.1 - - [16/Mar/2024 12:10:13] "POST / HTTP/1.1" 200 -

in the terminal when Im running a Dash application.
```
server = flask.Flask(__name__)
app = dash.Dash(
    __name__,
    server=server,
    external_stylesheets=[dbc.themes.BOOTSTRAP],
    meta_tags=[
        {"name": "viewport", "content": "width=device-width, initial-scale=1"}
    ],
)
```
(https://github.com/AlexKurek/srt-py/blob/master/srt/dashboard/app.py#L53)

How do I silence them? I have found a lot of methods, e.g. [Disable console messages in Flask server](https://stackoverflow.com/questions/14888799/disable-console-messages-in-flask-server), but none of them worked. My guess is that since they worked in previous versions of Flask - could this be a regression?

Environment:

- Python version: 3.11
- Flask version: 3.0.2

**Comment by codespearhead** [comment]
Try [this](https://stackoverflow.com/a/30086809).

**Reply by AlexKurek** [reply]
@codespearhead I did:
```
    server = flask.Flask(__name__) # these messages "127.0.0.1 - - [16/Mar/2024 12:10:13] "POST / HTTP/1.1" 200 -"" are generated by Flask
    app = dash.Dash(
        __name__,
        server=server,
        external_stylesheets=[dbc.themes.BOOTSTRAP],
        meta_tags=[
            {"name": "viewport", "content": "width=device-width, initial-scale=1"}
        ],
    )
    app.logger.setLevel(logging.ERROR)
    server.logger.setLevel(logging.ERROR)
```
and there is no change, I still get these messages.
In general, this is the most recommended thing to do on the internet. People report that it worked in earlier versions of Flask.

**Reply by codespearhead** [reply]
Help us reproduce the error on our machines with a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example#:~:text=In%20computing%2C%20a%20minimal%20reproducible,to%20be%20demonstrated%20and%20reproduced.). This is what I got so far, but the errors I'm getting are unrelated to what's being reported.

```bash
mkdir myproject &&
cd myproject &&
python3 -m venv .venv &&
source .venv/bin/activate &&
pip install flask dash
```

```bash
cat > main.py <<'EOF'
server = flask.Flask(__name__)
app = dash.Dash(
    __name__,
    server=server,
    external_stylesheets=[dbc.themes.BOOTSTRAP],
    meta_tags=[
        {"name": "viewport", "content": "width=device-width, initial-scale=1"}
    ],
)
EOF
```

```bash
python main.py
```

**Comment by jtrip** [comment]
Perhaps it is the application server that is doing the logging, which wsgi server are you using? Check to see how to limit logging there.

**Reply by AlexKurek** [reply]
Its `waitress`, I found `from waitress import serve`.

**Comment by jtrip** [comment]
Great, have you tried to silence the logging at that level to see if that resolves the issue for you?

I've seen some related discussion here:
https://github.com/Pylons/waitress/issues/151

and in the documentation:
https://docs.pylonsproject.org/projects/waitress/en/stable/logging.html

they suggest something like `setup_console_handler = False` as a parameter in `serve()` 
but I've also seen something like:
`serve(app, _quiet=True)`
not sure if there is a version difference there or if either might work.


Edit: 
Oh I also see:
"That _quiet parameter is a testing shim, and is not public API."
So that is not recommended, but one of the other solutions in that issue (151) should help.

**Reply by AlexKurek** [reply]
Not working:
`logging.getLogger('waitress').setLevel(logging.ERROR)`
and
`logging.getLogger().addHandler(logging.NullHandler())`
and
`serve(app_server, _quiet=True)`

**Comment by jtrip** [comment]
Can you provide a minimum reproducible example? 
I don't know much about dash (maybe that is also doing some logging?) not sure how to set that up or how the waitress serve call integrates with the example in your original post.

Also it might be good to analyze the results of the things you've tried beyond does or does not work. The first solution you tried is setting the log level, does the app seem to be respecting that or are you seeing entries on a lower level?

**Reply by AlexKurek** [reply]
> Can you provide a minimum reproducible example?

I will be difficult, since its not my code and its rather not small: https://github.com/MITHaystack/srt-py/issues/19. Also its not trivial to install and run.

I dont think any other output is/was from dash.

**Comment by jtrip** [comment]
Ok yeah large applications can be a challenge when debugging.

I just made a simple flask + dash + waitress app. Like this:

```
from flask import Flask
from dash import Dash
from dash import html
from waitress import serve
import logging

waitress_logger = logging.getLogger('waitress')
waitress_logger.setLevel(logging.ERROR)

flask_app = Flask(__name__)
flask_app.logger.handlers = waitress_logger.handlers

app = Dash(__name__, server=flask_app, routes_pathname_prefix='/dash/')
app.layout = html.Div(children=[
    html.H1(children='Hello Dash'),
])

@flask_app.route('/')
def hello():
    return 'Hello, this is the Flask part! Navigate to /dash/ for the Dash app.'


if __name__ == '__main__':
    serve(flask_app, host='0.0.0.0', port=8080)
 
```

and I noticed that without the logging handlers setting the level properly I was getting INFO level messages about a task queue.

However with the handlers added and level set like in this example I no longer see INFO messages, but would expect to see ERROR messages.

So I think that should work. Maybe you could share more details about the exact messages are trying to silence.

**Reply by AlexKurek** [reply]
It was easy to silence the queue messages: https://github.com/AlexKurek/srt-py/commit/315ec8c88516114d11c808410fbb8ed981e5c0bb.
I added:
```
  waitress_logger = logging.getLogger('waitress')
  waitress_logger.setLevel(logging.ERROR)
  server.logger.handlers = waitress_logger.handlers
  app.logger.handlers = waitress_logger.handlers
```
but:
> 127.0.0.1 - - [20/Mar/2024 16:47:25] "POST / HTTP/1.1" 200 -
> 127.0.0.1 - - [20/Mar/2024 16:47:25] "POST / HTTP/1.1" 200 -
> 127.0.0.1 - - [20/Mar/2024 16:47:26] "POST / HTTP/1.1" 200 -
> 127.0.0.1 - - [20/Mar/2024 16:47:26] "POST / HTTP/1.1" 200 -
> 127.0.0.1 - - [20/Mar/2024 16:47:26] "POST / HTTP/1.1" 200 -
> [...]

is still there.

**Reply by jtrip** [reply]
Interesting, when I expand my example so the endpoints accept POST and use curl to POST to them I do not get anything in the console.

but that is because I am letting waitress do the serving in the conditional at the bottom, and running this via `python main.py`, are you by chance running this via the flask dev server instead? like `flask run` etc.?

**Reply by jtrip** [reply]
if so, that can be disabled with the werkzeug logger

```
werkzeug_logger = logging.getLogger('werkzeug')
werkzeug_logger.disabled = True
```

**Reply by AlexKurek** [reply]
@jtrip I dont think so, in the code I only see `import flask`.

> werkzeug_logger = logging.getLogger('werkzeug')
> werkzeug_logger.disabled = True

It didnt work :(

**Comment by AlexKurek** [comment]
If you look at some threads, you can see that more and more methods have stopped working over the years. For example:
https://stackoverflow.com/a/74137152

**Reply by jtrip** [reply]
I can't see how you are drawing that conclusion from that stack overflow post.

As I mentioned logging configuration seems to be working perfectly well in my example using either Waitress or the Flask dev wekzeug server. It is possible there is some other component in your application that is not using that configuration I suppose but honestly I suspect maybe your implementation of the logging configuration is not set early enough in the application or it is not using the correct server.

Can you please confirm how you are running your application? Specifically, what command do you run to start it?

**Reply by AlexKurek** [reply]
@jtrip This whole application is started by running `srt_runner.py` and Dashboard is launched from there (https://github.com/MITHaystack/srt-py/blob/master/bin/srt_runner.py#L23).

Everything I tried, I placed here: https://github.com/MITHaystack/srt-py/blob/master/srt/dashboard/app.py#L63

**Reply by jtrip** [reply]
Thanks Alex, I looked into the project a little more but the setup is a little complex. 

I tried to get the project running but ran into a conda build error, I saw your steps here and tried to follow.
https://github.com/MITHaystack/srt-py/issues/19#issuecomment-1970720214

However at the end of that the srt-py project has been removed `rm -rf ~/srt-py/` and also when I tried other steps I noticed some python dependencies were not installed by conda, not sure if there is another step there that I am missing.

If you can give me some more details on the setup process I will try to get it running and try a few possible solutions.

**Reply by AlexKurek** [reply]
This `rm -rf ~/srt-py/` simply removes the cloned repo. Before that, the project is installed in conda with `conda install -y --use-local srt-py`.
Unfortunately, you need a serial to USB adapter at /dev/ttyUSB0 and a TV dongle that is RTL-SDR compatible (https://www.rtl-sdr.com/buy-rtl-sdr-dvb-t-dongles/). The project will not run without it.

**Reply by jtrip** [reply]
ah ok yeah, I do have an RTL SDR, but was trying to set it up in a cloud vm, so I might try to find a physical machine I can use. I'm guessinh this probably won't run on raspberry pi although maybe, let me know if you know.

Otherwise I think you were on the right track, that run_srt_dashboard function is using waitress so I think that is the right way to go in terms of `getlogger()`, however I think those request log entries might not be subject to log levels (just guessing) maybe try using the disabled attribute like I did for werkzeug.

maybe it would look something like this:
```
  waitress_logger = logging.getLogger('waitress')
  waitress_logger.disabled = True
  server.logger.handlers = waitress_logger.handlers
  app.logger.handlers = waitress_logger.handlers
```

**Reply by AlexKurek** [reply]
I pasted it here:
https://github.com/MITHaystack/srt-py/blob/master/srt/dashboard/app.py#L63
but it did not work. I did that too:
```
  waitress_logger = logging.getLogger('waitress')
  waitress_logger.disabled = True
  app_server.logger.handlers = waitress_logger.handlers
```
here: https://github.com/MITHaystack/srt-py/blob/master/bin/srt_runner.py#L32
but this did not help either.

**Reply by AlexKurek** [reply]
This could be important:
https://stackoverflow.com/a/57086684/6764984
> Flask apparently no longer uses logging for logging, and has switched to the [click](https://click.palletsprojects.com/en/7.x/) package

I pasted
```
log = logging.getLogger('werkzeug')
log.setLevel(logging.ERROR)

def secho(text, file=None, nl=None, err=None, color=None, **styles):
    pass

def echo(text, file=None, nl=None, err=None, color=None, **styles):
    pass

click.echo = echo
click.secho = secho
```
here: https://github.com/MITHaystack/srt-py/blob/master/srt/dashboard/app.py#L63 but also no result.

**Reply by jtrip** [reply]
I don't believe that would be important as it actually seems to be a misunderstanding or some confusion about what messages on the console are a flask app doing some logging and a flask/click command. Flask uses the standard python `logging` library as the documentation accurately describes and you can see in the flask source code.

https://flask.palletsprojects.com/en/3.0.x/logging/
https://github.com/pallets/flask/blob/b90a4f1f4a370e92054b9cc9db0efcb864f87ebe/src/flask/logging.py

I did notice this section that might be worth looking into:
https://flask.palletsprojects.com/en/3.0.x/logging/#removing-the-default-handler

Otherwise it might be worth looking into Dash to see if it might be creating a logger.

**Reply by AlexKurek** [reply]
I pasted:
```
app.logger.removeHandler(default_handler)
server.logger.removeHandler(default_handler)
```
here: https://github.com/MITHaystack/srt-py/blob/master/srt/dashboard/app.py#L63
but it didnt work. What do you mean by looking into Dash?

**Reply by jtrip** [reply]
Dash might be doing some logging, look to see if they document anything about working with logging.

You may also need to do more debugging, just trying to paste in some code will not help you understand what is going on.
Also maybe you have some other configuration impacting the application at runtime.
